packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -123,19 +177,92 @@ class DatePickerInput extends React.Component {
     }
   }
 
-  handleOnChange(event) {
-    if (!DateUtil.validDateInput(event.target.value)) {
-      return;
+  handleOnKeyDown(event) {
+    if (this.props.onKeyDown) {
+      this.props.onKeyDown(event);
     }
+  }
 
-    if (this.props.onChange) {
-      this.props.onChange(event);
+  handleDayChange(event) {
+    const inputValue = event.target.value.replace(/\D/gm, '');
+    this.handleDateChange(event, DateUtil.inputType.DAY, inputValue);
+  }
+
+  handleMonthChange(event) {
+    const inputValue = event.target.value.replace(/\D/gm, '');
+    this.handleDateChange(event, DateUtil.inputType.MONTH, inputValue);
+  }
+
+  handleYearChange(event) {
+    const inputValue = event.target.value.replace(/\D/gm, '');
+    this.handleDateChange(event, DateUtil.inputType.YEAR, inputValue);
+  }
+
+  handleOnFocus(event) {
+    this.setState({ isFocused: true });
+    if (this.props.onFocus) {
+      this.props.onFocus(event);
     }
   }
 
-  handleOnKeyDown(event) {
-    if (this.props.onKeyDown) {
-      this.props.onKeyDown(event);
+  handleOnBlur(event) {
+    this.setState({ isFocused: false });
+
+    if (this.props.onBlur) {
+      this.props.onBlur(event);
+    }
+  }
+
+  handleDateChange(event, type, value) {
+    if (type === DateUtil.inputType.DAY) {
+      this.setState({
+        day: value,
+      });
+    } else if (type === DateUtil.inputType.MONTH) {
+      this.setState({
+        month: value,
+      });
+    } else if (type === DateUtil.inputType.YEAR) {
+      this.setState({
+        year: value,
+      });
+    }
+
+    this.moveFocusOnChange(value, type, this.variant);
+
+    /**
+     * Sets the day, month and year based on input values, formats them
+     * based on the date format variant, and passes the formatted date to onChange.
+     */
+    const day = type === DateUtil.inputType.DAY ? value : this.state.day;
+    const month = type === DateUtil.inputType.MONTH ? value : this.state.month;
+    const year = type === DateUtil.inputType.YEAR ? value : this.state.year;
+    const dateFormat = DateUtil.getFormatByLocale(this.props.intl.locale);
+
+    let tempDate;
+    let formattedDate;
+
+    if (day && month && year) {
+      if (this.variant === 'YYYY-MM-DD') {
+        if (day.length === 2) {
+          tempDate = DateUtil.formatISODate(year.concat('-', month)
+            .concat('-', day), 'YYYY-MM-DD');
+          formattedDate = DateUtil.strictFormatISODate(tempDate, dateFormat);
+        }
+      } else if (year.length === 4) {
+        tempDate = DateUtil.formatISODate(year.concat('-', month)
+          .concat('-', day), 'YYYY-MM-DD');
+        formattedDate = DateUtil.strictFormatISODate(tempDate, dateFormat);
+      }
+    }
+
+    // Allows easy deletion of input values.
+    if (day === '' && month === '' && year === '') {
+      this.props.onChange(event, '');
```

Comment:
At first glance, I thought this would call onChange twice (once here and another on line 265). Could we move this to an `else if` block for readability?

Reply:
Done! [0b496eb](https://github.com/cerner/terra-framework/pull/1004/commits/0b496eb729d1b0fc5851cc50d820b24a267a67ec)